Add bonding and vlan tagging support to Bottlerocket (#2596)
bottlerocket-os/bottlerocket: PROVISIONING-METAL.md

#### `net.toml` structure

The configuration file must be valid TOML and have the filename `net.toml`.

> The first and required top level key in the file is `version`; the latest is version `3`.

Reviewer: Since we've created several `version`s of `net.toml` rapidly would it be helpful have a table that maps `version` with the Bottlerocket version?

I think this could be a little confusing otherwise especially if someone is only a minor version back.

Author: I'll add in a table for the releases that enabled this. I was able to avoid making a breaking change so we don't have any required work for users upgrading for now. Nonetheless, as we iterate this is going to be more and more important to track.

  * `via` (IP address): Gateway IP address.  If no gateway is provided, a scope of `link` is assumed.
  * `route-metric` (integer): Relative route priority.

> Version `3` adds in support for bonding and vlan tagging. The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt). Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces. Physical devices like ethernet adapters are considered "interfaces", and network devices encompass other kinds of devices like bonds and vlans. This is a *breaking change* from version 2 and requires the user to specify 

Reviewer 2: Can you break these up to be one sentence per line?

Reviewer 2: Also, looks like you got interupted.

> and requires the user to specify...

Author: Whoops, I totally forgot to update this. We actually don't have a breaking change, which is great! So I'll update the docs to be accurate. I'll also look at cleaning up the markdown more thoroughly as well.

> Migrating to version `3` from version `2` requires changes to how devices are defined. For example, if `[eno1]` is the current device name, changing a config from `[eno1]` to `[interface.eno1]` is all that is required. 

Reviewer: Is anything special required for version `1` to `3`?

Author: I edited this out, there is nothing specific a user must do to move from version `1` or `2` when moving to `3` (besides updating `version = 3`).

Bonding configuration is created via a network device (`device`):

> * `bond` (map): To create a `bond` network device, specify `kind = "bond"`.

Reviewer 2: If you're doing an update - please add a blank line before the start of the bullet list. There is another PR that is cleaning up a lot of markdown lint errors. After that merges we could potentially add a markdownlint CI job. If we do that, it will flag this kind of thing where there is no separation before lists and code blocks.

Author: I'll update according the markdownlint so its consistent.

    * `miimon-frequency` (int in ms): MII Monitoring frequency in miliseconds
    * `miimon-updelay` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
    * `arpmon-interval` (int in seconds): Number of seconds between intervals to determine link status, must be greater than 0
    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16

Vlan tagging is configured as a new network device (`device`) stacked on another device:
* `vlan` (map): Define by `kind = "vlan"`
  * `device` (string for device): Defines the device the vlan should be configured on. If VLAN tagging is required, this device should recieve all IP address configuration instead of the underlying device.
  * `id` (int): Number between 0 and 4096 specifying the vlan tag on the device

> 

Reviewer 2: nit: extra blank line

Again, just a markdown lint concern, should we choose to go that route.

Author: Will fix.

  * `via` (IP address): Gateway IP address.  If no gateway is provided, a scope of `link` is assumed.
  * `route-metric` (integer): Relative route priority.

Version `3` adds in support for bonding and vlan tagging. 
The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).

> Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.

Reviewer 3: > Version `3` also adds the concept of network devices vs network interfaces.

Is this still true in terms of config structure?

Author: It does, but I failed to explain what that means to the end user. `interface` really just means there is no `kind` specified. I'll add some extra detail around how one must treat interfaces differently than network devices. The other important one is you have to find the right name for an interface but can choose your own name for devices.

Author: I added a few more lines trying to explain that interfaces look like what came before and `virtual devices` are new and different.

> Physical devices like ethernet adapters are considered "interfaces", and network devices encompass other kinds of devices like bonds and vlans.

Reviewer 3: This sentence can probably be omitted.

Author: I kind of went the other way since naming comes into play, essentially you have to give us the exact name of a physical(pre-existing) device, and you have to choose a valid name for the others. So it becomes relevant to know which one you are dealing with.

The default type of device is an interface and the syntax is the same as previous versions.

> For virtual network devices, a `kind` is required.

Reviewer 3: Perhaps here is where we explain that if `kind` is missing, we assume the device is an interface.

Currently, `bond` and `vlan` are the two supported `kind`s.

> Bonding configuration is created via a network device (`device`):

Reviewer 3: This is no longer true. :)

Author: Good catch. I'll fix.

> * `bond` (map): To create a `bond` network device, specify `kind = "bond"`.

Reviewer 3: This is a little bit confusing to me - `bond` isn't a top level key or map.  Perhaps this should be some sort of heading like `Bond settings`?

Author: Yeah, it was at one point in my head, but it no longer is. I think changing the heading is a good call.

Author: I updated these lines, can you confirm they make more sense now?

  * `interfaces` (list of quoted strings of interfaces): Which interfaces should be added to the bond (i.e. `["eno1"]`). The first in the list is considered the default `primary`.
  * `mode` (string): Currently `active-backup` is the only supported option
  * `min-links` (int): Number of links required to bring up the device
  * `monitoring` (map): Value must all be of `miimon` or `arpmon` type. The user must choose one type of monitoring and configure it fully in order for the bond to properly function. See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.

>     * `miimon-frequency` (int in ms): MII Monitoring frequency in miliseconds

Reviewer 3: As discussed offline - let's add the time scale (`milliseconds`, etc) into the setting name to make it more obvious.

Author: I did a bit more reading on this and realized I mis-read the ARP setting being in seconds. [From what I can tell,](https://github.com/openSUSE/wicked/blob/master/schema/bonding.xml#L44) `wicked` under the hood expects everything in milliseconds. The only thing I've found that may have given me the impression it was in seconds is from netplan's docs which call out a specific field `ArpIntervalSec` property but the default is still milliseconds. In light of everything being in milliseconds, should we just adjust everything to document all time settings are in milliseconds or still continue with `miimon-frequency-milliseconds`?

Reviewer 3: I think it probably makes sense to include the time scale in the setting name to make it obvious to the user.

Author: I made it `-ms` for the postfix. I think this is a reasonable compromise.

    * `miimon-updelay` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
    * `arpmon-interval` (int in seconds): Number of seconds between intervals to determine link status, must be greater than 0
    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16

> Vlan tagging is configured as a new network device (`device`) stacked on another device:

Reviewer 3: We can probably remove the `device` referenced here right?

  * `via` (IP address): Gateway IP address.  If no gateway is provided, a scope of `link` is assumed.
  * `route-metric` (integer): Relative route priority.

Version `3` adds in support for bonding and vlan tagging. 
The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).

> Future support may include other bonding options - pull requests are welcome! Version `3` also adds the concept of network devices vs network interfaces.

Reviewer 4: nit: style consistency

```suggestion
Future support may include other bonding options - pull requests are welcome!
Version `3` also adds the concept of network devices vs network interfaces.
```

Author: I'll fix it.

The default type of device is an interface and the syntax is the same as previous versions.
If no `kind` is specified, it is assumed to be an interface.
The name of an interface must match an existing interface on the system such as `eno1` or `enp0s16`.
For virtual network devices, a `kind` is required.
Currently, `bond` and `vlan` are the two supported `kind`s.
Virtual network devices are created, and therefore a name has to be chosen.
Names for virtual network devices must conform to [kernel naming requirements](https://elixir.bootlin.com/linux/v5.10.102/source/include/uapi/linux/if.h#L33).

Bonding configuration creates a virtual network device across several other devices:

* Bonding configuration (map):
  * `kind = "bond"`: This setting is required to specify a bond device.
  * `interfaces` (list of quoted strings of interfaces): Which interfaces should be added to the bond (i.e. `["eno1"]`). The first in the list is considered the default `primary`.
  * `mode` (string): Currently `active-backup` is the only supported option
  * `min-links` (int): Number of links required to bring up the device

>   * `monitoring` (map): Value must all be of `miimon` or `arpmon` type. The user must choose one type of monitoring and configure it fully in order for the bond to properly function. See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.

Reviewer 4: The main nit is grammar (Values), but also the long line could be broken up.

```suggestion
  * `monitoring` (map): Values must all be of `miimon` or `arpmon` type.
    The user must choose one type of monitoring and configure it fully in order for the bond to properly function.
    See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.
```

Author: I'll fix this as well.

    * `miimon-frequency-ms` (int in ms): MII Monitoring frequency in milliseconds
    * `miimon-updelay-ms` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
    * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0

>     * `arpmon-validate-ms` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link

Reviewer 4: Is this a milliseconds value?

```suggestion
    * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link
```

Author: I was too aggressive in my adding `-ms`. I'll pull them out, you are correct, the validate is not an `-ms`.

>     * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16

Reviewer 5: nit: maybe indent the miimon and arpmon options another level

Author: They are indented one, did you want more or for them to be inline with the rest of the options?

Reviewer 5: Not sure if it wasn't rendering right for me earlier, but you're right and it looks fine to me now.

>     * `arpmon-targets-ms` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16

Reviewer 4: Same here, am I not understanding what the `ms` suffix means? I thought it was milliseconds.

```suggestion
    * `arpmon-targets` (list of quoted IPv4 address including prefix): List of targets to use for validating ARP. Min = 1, Max = 16
```

Author: Same as above, was too aggressive.

  * `via` (IP address): Gateway IP address.  If no gateway is provided, a scope of `link` is assumed.
  * `route-metric` (integer): Relative route priority.

Version `3` adds in support for bonding and vlan tagging. 
The support is limited to mode `1` (`active-backup`) for [bonding](https://www.kernel.org/doc/Documentation/networking/bonding.txt).
Future support may include other bonding options - pull requests are welcome! 

> Version `3` also adds the concept of network devices vs network interfaces.

Reviewer 3: Here we say `network devices vs. network interfaces`, but the following sentence refers to an interface as a type of device.

Perhaps we should say something like `Version 3 adds the concept of virtual network devices in addition to interfaces`

The default type of device is an interface and the syntax is the same as previous versions.
The name of an interface must match an existing interface on the system such as `eno1` or `enp0s16`.
For virtual network devices, a `kind` is required.
If no `kind` is specified, it is assumed to be an interface.
Currently, `bond` and `vlan` are the two supported `kind`s.
Virtual network devices are created, and therefore a name has to be chosen.

> Names for virtual network devices must conform to [kernel naming requirements](https://elixir.bootlin.com/linux/v5.10.102/source/include/uapi/linux/if.h#L33).

Reviewer 3: This link isn't particularly helpful - perhaps we should enumerate the short list of requirements here?

Author: I'll add it to be more explicit instead.

* Bonding configuration (map):
  * `kind = "bond"`: This setting is required to specify a bond device. Required.
  * `interfaces` (list of quoted strings of interfaces): Which interfaces should be added to the bond (i.e. `["eno1"]`). The first in the list is considered the default `primary`. These interfaces are "consumed" so no other configuration can refer to them. Required.
  * `mode` (string): Currently `active-backup` is the only supported option. Required.
  * `min-links` (int): Number of links required to bring up the device
  * `monitoring` (map): Values m ust all be of `miimon` or `arpmon` type. 
    The user must choose one type of monitoring and configure it fully in order for the bond to properly function. 
    See [section 7](https://www.kernel.org/doc/Documentation/networking/bonding.txt) for more background on what to choose.
    * `miimon-frequency-ms` (int in ms): MII Monitoring frequency in milliseconds
    * `miimon-updelay-ms` (int in ms): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay-ms` (int in ms): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds

>     * `arpmon-interval-ms` (int in ms): Number of milliseconds between intervals to determine link status, must be greater than 0

Reviewer 5: "(int in ms)" is hard for me to parse, like it implies there's some sort of numeric conversion that has to happen.

```suggestion
    * `miimon-frequency-ms` (integer): MII Monitoring frequency in milliseconds
    * `miimon-updelay-ms` (integer): MII Monitoring delay before the link is enabled after link is detected in milliseconds
    * `miimon-downdelay-ms` (integer): MII Monitoring delay before the link is disabled after link is no longer detected in milliseconds
    * `arpmon-interval-ms` (integer): Number of milliseconds between intervals to determine link status, must be greater than 0
```

>     * `arpmon-validate` (enum of `all`, `none`, `active`, `backup`): What packets should be used to validate link

Reviewer 5: nit: since this is a TOML file, it might be better to only specify types and values that TOML understands

```suggestion
    * `arpmon-validate` (one of `"all"`, `"none"`, `"active"`, `"backup"`): What packets should be used to validate link
```

to = "10.10.10.0/24"
from = "192.168.14.5"
via = "192.168.14.25"

# Creating a bond and using vlan tagging with comments:

# A bond is a `device` that is of `type` `bond`
[bond0]
kind = "bond"
# Currently `active-backup` is the only supported option
mode = "active-backup"
min-links = 2
# In this case, the vlan will have addressing, the bond is simply there for use in the vlan
dhcp4 = false
dhcp6 = false
# The first interface in the array is considered `primary` by default
interfaces = ["eno1", "eno2"]

[bond0.monitoring]
miimon-frequency = 100 # 100 ms
miimon-updelay = 200 # 200 ms
miimon-downdelay = 200 # 200 ms

# VLAN42 is the name of the device

> [devices.VLAN42]

Reviewer 3: Let's remove the `devices`

dhcp4 = true

[bond1.monitoring]
arpmon-interval = 2
arpmon-validate = "all"
arpmon-targets = ["192.168.1.1", "10.0.0.2"]
```

#### **An additional note on network device names**

Interface name policies are [specified in this file](https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/release/80-release.link#L6); with name precedence in the following order: onboard, slot, path.
Typically on-board devices are named `eno*`, hot-plug devices are named `ens*`, and if neither of those names are able to be generated, the “path” name is given, i.e `enp*s*f*`.

#### Networking configuration versions and Releases

Older networking configuration versions (such as `1` or `2`) are supported in newer releases. In order to use a newer version, the following table provides guidance on what release first enabled the version.

| Network Configuration Version | First Release |
| ----------------------------- | ------------- |
| Version 1                     | [v1.9.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.9.0)   |
| Version 2                     | [v1.10.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.10.0) |

> | Version 3                     | v.1.12.0 |

Reviewer 6: ```suggestion
| Version 3                     | v1.12.0 |
```
I wonder if it's best we left this as `Unreleased` since v1.12.0 doesn't exist yet and probably won't exist for a while. Then during the v1.12.0 release we'd have to remember to flip this back.

dhcp6 = true
```

#### **An additional note on network device names**

Interface name policies are [specified in this file](https://github.com/bottlerocket-os/bottlerocket/blob/develop/packages/release/80-release.link#L6); with name precedence in the following order: onboard, slot, path.
Typically on-board devices are named `eno*`, hot-plug devices are named `ens*`, and if neither of those names are able to be generated, the “path” name is given, i.e `enp*s*f*`.

#### Networking configuration versions and Releases

Older networking configuration versions (such as `1` or `2`) are supported in newer releases. In order to use a newer version, the following table provides guidance on what release first enabled the version.

| Network Configuration Version | First Release                                                                   |
|-------------------------------|---------------------------------------------------------------------------------|
| Version 1                     | [v1.9.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.9.0)   |
| Version 2                     | [v1.10.0](https://github.com/bottlerocket-os/bottlerocket/releases/tag/v1.10.0) |

> | Version 3                     | Unreleased                                                                      |

Reviewer 5: nit: it might be better to go ahead and optimistically set this to `1.12.0`, and just fix it up prior to merge if it doesn't make the release

Otherwise we might miss changing it in a follow-up PR prior to cutting the release.

Author: I originally had it that way. We had expected to merge this a while ago and have other edits in the repo since then and flip then. Since we are still in PR, it makes sense to just change it for the next release.

Author: I decided to leave this for now, I'll be tracking quite a bit for the next release and plan to make sure this makes it into the release changes for whatever release it does make.
